File: cpp/rotate_image.cc
/*
 * Filename: rotate_image.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sun 26 Jul 2015 12:57:54 PM EDT
 */

#include <iostream>
using namespace std;

class Solution {
	public:
		void rotate(vector<vector<int> >& matrix) {
			for (int i = 0, j = matrix.size() - 1; i < j; i++, j--)  {  
				for (int k = i, d = j; k < j; k++, d--) {  
					int t = matrix[i][k];  

					matrix[i][k] = matrix[d][i];  
					matrix[d][i] = matrix[j][d];  
					matrix[j][d] = matrix[k][j];  
					matrix[k][j] = t;  
				}  
			}
		}
};

File: java/RotateImage.java
/*
 * Filename: RotateImage.java
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Tue 12 May 2015 08:56:20 PM EDT
 */

public class Solution {
	public void rotate(int[][] matrix) {
		if (matrix == null || matrix.length == 0 || matrix[0].length == 0) {
			return null;
		}

		int n = matrix.length;
		for (int i = 0; i < n; i++) {
			for (int j = i + 1; j < n; j++) {
				int temp = matrix[i][j];
				matrix[i][j] = matrix[j][i];
				matrix[j][i] = temp;
			}
		}

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n / 2; j++) {
				int temp = matrix[i][j];
				matrix[i][j] = matrix[i][n-j-1];
				matrix[i][n-j-1] = temp;
			}
		}
	}
}
